source/server/listener_manager_impl.cc

```diff
@@ -69,57 +69,98 @@ ProdListenerComponentFactory::createFilterFactoryList_(
+    : Json::Validator(json, Json::Schema::LISTENER_SCHEMA), parent_(parent),
       address_(Network::Utility::resolveUrl(json.getString("address"))),
-      global_scope_(server.stats().createScope("")),
+      global_scope_(parent_.server_.stats().createScope("")),
       bind_to_port_(json.getBoolean("bind_to_port", true)),
       use_proxy_proto_(json.getBoolean("use_proxy_proto", false)),
       use_original_dst_(json.getBoolean("use_original_dst", false)),
       per_connection_buffer_limit_bytes_(
           json.getInteger("per_connection_buffer_limit_bytes", 1024 * 1024)),
-      listener_tag_(factory.nextListenerTag()) {
+      listener_tag_(parent_.factory_.nextListenerTag()), name_(name),
+      workers_started_(workers_started), hash_(hash) {
 
   // ':' is a reserved char in statsd. Do the translation here to avoid costly inline translations
   // later.
   std::string final_stat_name = fmt::format("listener.{}.", address_->asString());
   std::replace(final_stat_name.begin(), final_stat_name.end(), ':', '_');
-
-  listener_scope_ = server.stats().createScope(final_stat_name);
-  ENVOY_LOG(info, "  address={}", address_->asString());
+  listener_scope_ = parent_.server_.stats().createScope(final_stat_name);
 
   if (json.hasObject("ssl_context")) {
     Ssl::ContextConfigImpl context_config(*json.getObject("ssl_context"));
-    ssl_context_ =
-        server.sslContextManager().createSslServerContext(*listener_scope_, context_config);
+    ssl_context_ = parent_.server_.sslContextManager().createSslServerContext(*listener_scope_,
+                                                                              context_config);
   }
 
-  filter_factories_ = factory.createFilterFactoryList(json.getObjectArray("filters"), *this);
-  socket_ = factory.createListenSocket(address_, bind_to_port_);
+  filter_factories_ =
+      parent_.factory_.createFilterFactoryList(json.getObjectArray("filters"), *this);
+}
+
+ListenerImpl::~ListenerImpl() {
+  // The filter factories may have pending initialize actions (like in the case of RDS). Those
+  // actions will fire in the destructor to avoid blocking initial server startup. If we are using
+  // a local init manager we should block the notification from trying to move us from warming to
+  // active. This is done here explicitly by setting a boolean and then clearing the factory
+  // vector for clarity.
+  initialize_canceled_ = true;
+  filter_factories_.clear();
 }
 
 bool ListenerImpl::createFilterChain(Network::Connection& connection) {
   return Configuration::FilterChainUtility::buildFilterChain(connection, filter_factories_);
 }
 
+void ListenerImpl::infoLog(const std::string& message) {
+  ENVOY_LOG(info, "{}: name={}, hash={}, address={}", message, name_, hash_, address_->asString());
+}
+
+void ListenerImpl::initialize() {
+  // If workers have already started, we shift from using the global init manager to using a local
```

Comment:
It might be possible. I was trying to keep things consistent with how it works currently. I will add TODO to figure this out.